### PR TITLE
Ability to obtain payments from node

### PR DESCRIPTION
**Pull request by fiksn**

### CODE REVIEW

File: lightning/api_payments.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+				return
+			default:
+				resp, err := api.GetForwardingHistory(context.Background(), Pagination{
+					Num:    uint64(batchSize),
+					From:   &from,
+					To:     &to,
+					Offset: lastOffset,
+				})
+				if err != nil || resp == nil {
+					return
```
Comment (dcrystalj) [inline lightning/api_payments.go:39]: close(stopChan) and out chan ?

File: lightning/rest.go [hunk 2]
```diff
@@ -227,6 +229,85 @@ func (h *HTTPAPI) HTTPGetChanInfo(ctx context.Context, req *http.Request, trans
...
+
+	if err != nil {
+		return nil, fmt.Errorf("http request failed %s", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http got error %d", resp.StatusCode)
+	}
+
+	defer resp.Body.Close()
```
Comment (dcrystalj) [inline lightning/rest.go:299]: not sure but shouldn't this happen right after err != nill ?

File: lightning/api.go [hunk 3]
```diff
@@ -143,8 +145,134 @@ type NodeInfoAPIExtended struct {
...
+	From     *time.Time
+	To       *time.Time
+}
+
+// PaymentStatus enum
+type PaymentStatus int
+
+// PaymentStatus values
+const (
+	Unknown PaymentStatus = 0
```
[the remark this answers is not in the document]
Comment (fiksn) [reply lightning/api.go]: I know `iota` but this has to match the original enum, so I was explicit.

File: lightning/api_payments.go (hunk 1 above)
Comment (fiksn) [reply lightning/api_payments.go:39]: 👍

File: lightning/api_payments.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+	x := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
+	return x.AddDate(0, -1*monthsAgo, 0)
+}
+
+// IterateForwardingHistory - get an interator to obtain ForwardingEvent's, second parameter is stop channel
+func IterateForwardingHistory(api LightingAPICalls, from, to time.Time, batchSize uint) (<-chan ForwardingEvent, chan<- bool) {
+	lastOffset := uint64(0)
+
+	stopChan := make(chan bool, 1)
```
Comment (dcrystalj) [inline lightning/api_payments.go:23]: why dont we use context to stop this routine?

File: raw/fetcher.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	}
+	f.AgentAPI = agent
+
+	return f, nil
+}
+
+func makePermanent(err error) error {
+	st := status.Convert(err)
+	if st.Code() == codes.Unknown {
+		if strings.Contains(st.Message(), "ConditionalCheckFailedException") {
```
Comment (dcrystalj) [inline raw/fetcher.go:63]: I will make sure you will never receive this message.

File: agent/gen.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i bash -p bash protoc-gen-go protoc-gen-go-grpc grpc-tools
+
+### Generate go code from protobuf
+set -euo pipefail
+
+rm -f *.pb.go
```
Comment (dcrystalj) [inline agent/gen.sh:7]: why is this needed. it should be overwritten anyway ?

File: cmd/balance-agent/main.go [hunk 7]
```diff
@@ -303,6 +304,30 @@ func getApp() *cli.App {
 			Value:  "mainnet",
 			Hidden: true,
 		},
+		&cli.StringFlag{
+			Name:   "grpc-url",
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:308]: grpc-url is ambiguos. It can be url to node or backend api.

because we are data company my suggestion is: datastore-url

File: cmd/balance-agent/main.go [hunk 8]
```diff
@@ -303,6 +304,30 @@ func getApp() *cli.App {
...
+			Usage:  "GRPC URL",
+			Value:  "",
+			Hidden: true,
+		},
+		&cli.Int64Flag{
+			Name:   "fetch-invoices",
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:314]: also ambiguous. It can potentially be fetch from datastore or from LN node. But I don't have better suggestion :D

File: cmd/balance-agent/main.go [hunk 9]
```diff
@@ -570,6 +597,55 @@ func nodeDataChecker(ctx *cli.Context) error {
 	return nil
 }
 
+func fetcher(ctx *cli.Context, apiKey string) {
+	// run with --grpc-url agent-api-staging-new.bolt.observer:443 --fetch-invoices 1 --fetch-forwards 1 --fetch-payments 1
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:601]: you can have this in "help" command. Remove staging url

File: cmd/balance-agent/main.go [hunk 10]
```diff
@@ -570,6 +597,55 @@ func nodeDataChecker(ctx *cli.Context) error {
...
+			glog.Warningf("GRPC get ligtning failure\n")
+			return
+		}
+		f, err := raw.MakeFetcher(apiKey, ctx.String("grpc-url"), itf)
+		if err != nil {
+			glog.Warningf("GRPC get fetcher failure %v\n", err)
+			return
+		}
+
+		val := int64(0)
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:616]: Bad name. What val?

File: cmd/balance-agent/main.go [hunk 11]
```diff
@@ -570,6 +597,55 @@ func nodeDataChecker(ctx *cli.Context) error {
 	return nil
 }
 
+func fetcher(ctx *cli.Context, apiKey string) {
+	// run with --grpc-url agent-api-staging-new.bolt.observer:443 --fetch-invoices 1 --fetch-forwards 1 --fetch-payments 1
+	if ctx.String("grpc-url") != "" && apiKey != "" {
+		glog.Infof("GRPC server URL: %s", ctx.String("grpc-url"))
+
+		itf := mkGetLndAPI(ctx)()
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:605]: what `itf` stands for?

File: cmd/balance-agent/main.go [hunk 12]
```diff
@@ -570,6 +597,55 @@ func nodeDataChecker(ctx *cli.Context) error {
...
+		f, err := raw.MakeFetcher(apiKey, ctx.String("grpc-url"), itf)
+		if err != nil {
+			glog.Warningf("GRPC get fetcher failure %v\n", err)
+			return
+		}
+
+		val := int64(0)
+		var t time.Time
+
+		// 0 means turn off the fetcher, negative values means report from that exact absolute unix time (i.e., -1 means report from unix time 1), positive value
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:619]: very complex. Can't you save structs in context ?

File: cmd/balance-agent/main.go [hunk 13]
```diff
@@ -570,6 +597,55 @@ func nodeDataChecker(ctx *cli.Context) error {
...
+		// try to report from that unix time (but if server says you already got something update timestamp)
+		val = ctx.Int64("fetch-invoices")
+		absVal := int64(0)
+		if val < 0 {
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:623]: how come this logic applies only to invoices? looks like you use invoices `val` logic for forwards and payments after line 634

File: raw/fetcher.go [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+		LightningAPI: l,
+	}
+
+	info, err := l.GetInfo(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	typ := 0
+	switch l.GetAPIType() {
```
Comment (dcrystalj) [inline raw/fetcher.go:40]: why do you need this switch. you can just cast it to int.

File: raw/fetcher.go [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+// MakeFetcher creates a new fetcher
+func MakeFetcher(authToken string, endpoint string, l api.LightingAPICalls) (*Fetcher, error) {
+	f := &Fetcher{
+		AuthToken:    authToken,
+		LightningAPI: l,
+	}
+
+	info, err := l.GetInfo(context.Background())
```
Comment (dcrystalj) [inline raw/fetcher.go:34]: context must come from main

File: raw/agent.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to dial %s", err.Error())
+	}
+
+	return conn, nil
+
+}
+
+// GetAgentAPI retrieves the go API
+func GetAgentAPI(endpoint, pubKey, authToken string) api.AgentAPIClient {
```
Comment (dcrystalj) [inline raw/agent.go:82]: not used outside of this pacakge so can be private

File: raw/agent.go [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+	itf, err := getConnection(endpoint, pubKey, authToken)
+	if err != nil {
+		fmt.Printf("Error %+v\n", err)
```
Comment (dcrystalj) [inline raw/agent.go:85]: lets just return errors to main and print errors there.

File: raw/fetcher.go [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+		if strings.Contains(st.Message(), "ConditionalCheckFailedException") {
+			return backoff.Permanent(err)
+		}
+	}
+
+	return err
+}
+
+// FetchInvoices will fetch and report invoices
+func (f *Fetcher) FetchInvoices(ctx context.Context, updateTimeWithLast bool, from time.Time) {
```
Comment (dcrystalj) [inline raw/fetcher.go:72]: `updateTimeWithLast` with last what? 
Maybe explain more in the comment. I suggest `shouldUpdateTimeToLatest`

File: raw/fetcher.go [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+			glog.V(2).Infof("Latest invoice timestamp: %v", t)
+			if t.After(from) {
+				from = t
+			}
+		}
+	}
+
+	outchan := GetInvoicesChannel(ctx, f.LightningAPI, from)
+
+	num := 0
```
Comment (dcrystalj) [inline raw/fetcher.go:94]: Its type tells you its number.

I suggest `cnt`

File: raw/fetcher.go [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+		select {
+		case <-ctx.Done():
+			return
+		case invoice := <-outchan:
+			data := &agent.DataRequest{
+				Timestamp: invoice.Timestamp.UnixNano(),
+				Data:      string(invoice.Message),
+			}
+
+			backoff.RetryNotify(func() error {
```
Comment (dcrystalj) [inline raw/fetcher.go:109]: if we cannot send data we should stop this for loop. client must wait until backend is ready otherwise we might have gaps.

You are not handling error returned from RetryNotify

File: lightning/api.go [hunk 21]
```diff
@@ -146,8 +149,215 @@ type NodeInfoAPIExtended struct {
...
 }
 
-// LightningAPI - generic API settings
-type LightningAPI struct {
+////////////////////////////////////////////////////////////////
+
+// Pagination struct
+type Pagination struct {
+	Offset   uint64 // Exclusive thus 1 means start from 2 (0 will start from beginning)
+	Num      uint64 // limit is 10k or so
```
Comment (dcrystalj) [inline lightning/api.go:157]: Not descriptive. I suggest Page

File: lightning/api.go [hunk 22]
```diff
@@ -146,8 +149,215 @@ type NodeInfoAPIExtended struct {
...
+	From     *time.Time
+	To       *time.Time
+}
+
+// PaymentStatus enum
+type PaymentStatus int
+
+// PaymentStatus values
+const (
+	PaymentUnknown PaymentStatus = 0
```
Comment (dcrystalj) [inline lightning/api.go:168]: why not `... = iota` ?

File: lightning/api.go [hunk 23]
```diff
@@ -146,8 +149,215 @@ type NodeInfoAPIExtended struct {
...
+type PaymentsResponse struct {
+	Payments []Payment
+	ResponsePagination
+}
+
+// RawMessage struct
+type RawMessage struct {
+	Timestamp      time.Time `json:"timestamp"`
+	Implementation string    `json:"implementation,omitempty"`
+
```
Comment (dcrystalj) [inline lightning/api.go:337]: remove empty line

File: lightning/api_clnsocket.go [hunk 24]
```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
 	return result, nil
 }
 
+// SubscribeForwards - API call
+func (l *ClnSocketLightningAPI) SubscribeForwards(ctx context.Context, since time.Time, batchSize uint16) (<-chan []ForwardingEvent, <-chan ErrorData) {
+	var reply ClnForwardEntries
+
+	if batchSize == 0 {
+		batchSize = 50
```
Comment (dcrystalj) [inline lightning/api_clnsocket.go:437]: create default batchsize constant at the top of file

Comment (dcrystalj) [inline lightning/api_clnsocket.go:437]: magic number

File: lightning/api_clnsocket.go [hunk 25]
```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
...
+	}
+
+	const maxErrors = 5
```
Comment (dcrystalj) [inline lightning/api_clnsocket.go:440]: magic number. why is this not parameter as well ?

File: cmd/balance-agent/main.go [hunk 26]
```diff
@@ -570,6 +597,55 @@ func nodeDataChecker(ctx *cli.Context) error {
...
+			absVal = -val
+		} else {
+			absVal = val
+		}
+
+		if val != 0 {
+			t = time.Unix(absVal, 0)
+			glog.Infof("Fetching invoices after %v\n", t)
+			go f.FetchInvoices(context.Background(), val > 0, t)
```
Comment (dcrystalj) [inline cmd/balance-agent/main.go:632]: I know this goroutines should never stop or error, but if they do it would be nice if we could return that error and print it here.

File: lightning/api_clnsocket.go [hunk 27]
```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
...
+			if err != nil {
+				glog.Warningf("Error getting forwards %v\n", err)
+				if errors >= maxErrors {
+					errorChan <- ErrorData{Error: err, IsStillRunning: false}
+					return
+				}
+				errorChan <- ErrorData{Error: err, IsStillRunning: true}
+			}
+
+			for _, one := range reply.Entries {
```
Comment (dcrystalj) [inline lightning/api_clnsocket.go:469]: if CallWithTimeout returns error. reply.Entries might be from previous loop?

I suggest we move `var reply ClnForwardEntries` on line 458 just in case

File: lightning/api_clnsocket.go [hunk 28]
```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
...
+					batch = make([]ForwardingEvent, 0, batchSize)
+				}
+			}
+
+			if len(batch) > 0 {
+				outChan <- batch
+				batch = make([]ForwardingEvent, 0, batchSize)
+			}
+
+			minTime = time.Time(reply.Entries[len(reply.Entries)-1].ReceivedTime).Add(-1 * time.Second)
```
Comment (dcrystalj) [inline lightning/api_clnsocket.go:509]: magic number. why do we need subtract one second ?

File: lightning/api_clnsocket.go [hunk 29]
```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
...
+		}
+	}()
+
+	return outChan, errorChan
+}
+
+// GetInvoicesRaw - API call
+func (l *ClnSocketLightningAPI) GetInvoicesRaw(ctx context.Context, pendingOnly bool, pagination RawPagination) ([]RawMessage, *ResponseRawPagination, error) {
```
Comment (dcrystalj) [inline lightning/api_clnsocket.go:517]: I see no unit tests for this function.

File: lightning/api_clnsocket.go [hunk 30]
```diff
@@ -425,6 +429,259 @@ func (l *ClnSocketLightningAPI) GetNodeInfoFull(ctx context.Context, channels bo
...
+	}
+
+	respPagination.FirstTime = minTime
+	respPagination.LastTime = maxTime
+
+	return ret, respPagination, nil
+}
+
+// GetForwardsRaw - API call
+func (l *ClnSocketLightningAPI) GetForwardsRaw(ctx context.Context, pagination RawPagination) ([]RawMessage, *ResponseRawPagination, error) {
```
Comment (dcrystalj) [inline lightning/api_clnsocket.go:623]: most of the logic is duplicated for all types of payments. please extract common logic out and unit test it.

File: lightning/api_grpc.go [hunk 31]
```diff
@@ -218,3 +225,509 @@ func (l *LndGrpcLightningAPI) GetChanInfo(ctx context.Context, chanID uint64) (*
 	ret := l.convertChan(resp)
 	return &ret, nil
 }
+
+// SubscribeForwards API
+func (l *LndGrpcLightningAPI) SubscribeForwards(ctx context.Context, since time.Time, batchSize uint16) (<-chan []ForwardingEvent, <-chan ErrorData) {
+	// We will first try obtaining ForwadingHistory and then move to SubscribeHtlc
+	const maxErrors = 5
+	sleepTime := 5 * time.Second
```
Comment (dcrystalj) [inline lightning/api_grpc.go:233]: magic numbers here

File: lightning/api_grpc.go [hunk 32]
```diff
@@ -218,3 +225,509 @@ func (l *LndGrpcLightningAPI) GetChanInfo(ctx context.Context, chanID uint64) (*
...
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// Do nothing
+			}
+			resp, err := l.Client.ForwardingHistory(ctx, req)
+
+			if err != nil {
+				glog.Warningf("Error getting ForwadingHistory %v\n", err)
```
Comment (dcrystalj) [inline lightning/api_grpc.go:264]: why do we need logging if we are channeling errors on errorChan. Wouldnt these errors be logged on the other end ?

File: lightning/api_grpc.go [hunk 33]
```diff
@@ -218,3 +225,509 @@ func (l *LndGrpcLightningAPI) GetChanInfo(ctx context.Context, chanID uint64) (*
...
+					time.Sleep(sleepTime)
+					errors++
+					continue
+				}
+
+				if event.EventType != routerrpc.HtlcEvent_FORWARD {
+					// Ignore non-forward events
+					continue
+				}
+
```
Comment (dcrystalj) [inline lightning/api_grpc.go:355]: this go routine is over complex. please split logic in sane size. Maybe try to write tests for it first and you will see its too complex.

File: lightning/api_grpc.go [hunk 34]
```diff
@@ -218,3 +225,509 @@ func (l *LndGrpcLightningAPI) GetChanInfo(ctx context.Context, chanID uint64) (*
...
+	respPagination.LastOffsetIndex = uint64(resp.LastOffsetIndex)
+	respPagination.FirstOffsetIndex = 0
+
+	ret := make([]RawMessage, 0, len(resp.ForwardingEvents))
+
+	minTime := time.Unix(1<<63-1, 0)
+	maxTime := time.Unix(0, 0)
+
+	for _, forwarding := range resp.ForwardingEvents {
```
Comment (dcrystalj) [inline lightning/api_grpc.go:432]: I have seen this logic in api_clnsocket.go. Extract it out

File: lightning/rest.go [hunk 35]
```diff
@@ -227,6 +229,174 @@ func (h *HTTPAPI) HTTPGetChanInfo(ctx context.Context, req *http.Request, trans
...
+
+	resp, err := h.Do(req)
+
+	if err != nil {
+		return nil, fmt.Errorf("http request failed %s", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http got error %d", resp.StatusCode)
+	}
```
Comment (dcrystalj) [inline lightning/rest.go:348]: all the logic until here  is duplicated with HTTPForwardEvents(). Please extract it out

File: raw/payments.go [hunk 36]
```diff
@@ -0,0 +1,118 @@
...
+// beginOfThisMonth := GetFirstDay(0)
+// endOfThisMonth := GetFirstDay(-1).Add(-1 * time.Millisecond)
+func GetFirstDay(monthsAgo int) time.Time {
+	now := time.Now()
+
+	x := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
+	return x.AddDate(0, -1*monthsAgo, 0)
+}
+
+func paginator(ctx context.Context, itf api.LightingAPICalls, getData GetRawData, from time.Time, pageSize int, outchan chan api.RawMessage) error {
```
Comment (dcrystalj) [inline raw/payments.go:70]: im missing unit tests for paginator

**Review by dcrystalj — CHANGES_REQUESTED** [review]
I suggest we focus on lnd first to ship it faster

File: raw/fetcher.go (hunk 5 above)
Comment (fiksn) [reply raw/fetcher.go:63]: I think there will still be some errors that are permanent and should not be retried and all others that shouldn't be. Of course the logic will change.

File: agent/gen.sh (hunk 6 above)
Comment (fiksn) [reply agent/gen.sh:7]: 👍

File: raw/fetcher.go (hunk 20 above)
Comment (fiksn) [reply raw/fetcher.go]: Good catch

File: lightning/api_clnsocket.go (hunk 27 above)
Comment (fiksn) [reply lightning/api_clnsocket.go:468]: Solved in a different way

File: cmd/balance-agent/main.go (hunk 26 above)
Comment (fiksn) [reply cmd/balance-agent/main.go]: Did it another way

File: lightning/rest.go [hunk 37]
```diff
@@ -45,196 +45,120 @@ func (h *HTTPAPI) SetTransport(transport *http.Transport) {
...
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("http got error %d", resp.StatusCode)
-	}
-
-	defer resp.Body.Close()
-
-	decoder := json.NewDecoder(resp.Body)
-
-	err = decoder.Decode(&graph)
+	err = h.doRequest(req, &graph)
```
Comment (dcrystalj) [inline lightning/rest.go:105]: why not h.doGetRequest(url, &graph)

File: lightning/rest.go [hunk 38]
```diff
@@ -442,3 +337,26 @@ func (h *HTTPAPI) GetHTTPRequest(getData GetDataCall) (*http.Request, *http.Tran
 
 	return req, trans, nil
 }
+
+func (h *HTTPAPI) doRequest(req *http.Request, data any) error {
```
Comment (dcrystalj) [inline lightning/rest.go:341]: missing test?

File: lightning/rest.go (hunk 37 above)
Comment (fiksn) [reply lightning/rest.go:105]: It can be POST as well.

File: raw/fetcher.go [hunk 39]
```diff
@@ -105,148 +128,83 @@ func (f *Fetcher) FetchInvoices(ctx context.Context, shouldUpdateTimeToLatest bo
...
+	getChan := func(ctx context.Context, itf api.LightingAPICalls, from time.Time) <-chan api.RawMessage {
+		return GetForwardsChannel(ctx, itf, from)
+	}
 
-			cnt++
-			if cnt%10 == 0 {
-				glog.V(2).Infof("Reported %d forwards (last timestamp %v)", cnt, forward.Timestamp)
-			}
-		}
+	pushData := func(ctx context.Context, data *agent.DataRequest) error {
```
Comment (dcrystalj) [inline raw/fetcher.go:186]: I see no added value in wrapping this into new variables. getTime, getChan and pushData.

**Review by dcrystalj — APPROVED** [review]

File: raw/fetcher.go [hunk 40]
```diff
@@ -135,76 +139,75 @@ func (f *Fetcher) fetch(
...
-	f.fetch(ctx, "FetchInvoices", "invoices", getTime, getChan, pushData, shouldUpdateTimeToLatest, from)
+	method := fetcherMethod{
+		methodName: "FetchInvoices",
+		entityName: "invoices",
+		getTime: fetcherGetTime(func(ctx context.Context) (*agent.TimestampResponse, error) {
+			return f.AgentAPI.LatestInvoiceTimestamp(ctx, &agent.Empty{})
+		}),
+		getChan: fetcherGetChan(GetInvoicesChannel),
+		pushData: fetcherPushData(func(ctx context.Context, data *agent.DataRequest) error {
+			_, err := f.AgentAPI.Invoices(ctx, data)
```
Comment (dcrystalj) [inline raw/fetcher.go:171]: move this part in f.fetch()
